### PR TITLE
Use generic pubgrub incompatibility reason

### PR DESCRIPTION
**Pull request by konstin**
Pubgrub got a new feature where all unavailability is a custom, instead of the reasonless `UnavailableDependencies` and our custom `String` type previously (https://github.com/pubgrub-rs/pubgrub/pull/208). This PR introduces a `UnavailableReason` that tracks either an entire version being unusable, or a specific version. The error messages now also track this difference properly.

The pubgrub commit is our main rebased onto the merged https://github.com/pubgrub-rs/pubgrub/pull/208, i'll push `konsti/main-rebase-generic-reason` to `main` after checking for rebase problems.

### CODE REVIEW

File: crates/uv/tests/pip_install_scenarios.rs [hunk 1]
```diff
@@ -591,12 +591,12 @@ fn dependency_excludes_non_contiguous_range_of_compatible_versions() {
               package-c==1.0.0
               package-c==2.0.0
           and package-c==1.0.0 depends on package-a<2.0.0, we can conclude that package-c<2.0.0 depends on package-a<2.0.0.
-          And because package-c==2.0.0 depends on package-a>=3.0.0, we can conclude that all versions of package-c depend on one of:
+          And because package-c==2.0.0 depends on package-a>=3.0.0, we can conclude that package-c depends on one of:
```
Comment (zanieb) [inline crates/uv/tests/pip_install_scenarios.rs:594]: Could we consider this change separately? It doesn't seem related to this pull request.

File: crates/uv-resolver/src/resolver/mod.rs [hunk 2]
```diff
@@ -950,7 +949,7 @@ impl<
                             .or_default()
                             .insert(version.clone(), IncompletePackage::Offline);
                         return Ok(Dependencies::Unavailable(
-                            "network connectivity is disabled, but the metadata wasn't found in the cache"
+                            "has metadata that was not found in the cache, but network connectivity is disabled"
```
Comment (zanieb) [inline crates/uv-resolver/src/resolver/mod.rs:952]: This one doesn't seem like an improvement. Can you explain your thought here?

File: crates/uv-resolver/src/resolver/mod.rs [hunk 3]
```diff
@@ -363,38 +363,37 @@ impl<
...
-                                    UnavailablePackage::InvalidStructure(_) => {
-                                        "was found, but has an invalid format"
-                                    }
-                                })
-                        } else {
-                            None
+                    // Check if the decision was due to the package being unavailable
+                    if let PubGrubPackage::Package(ref package_name, _, _) = next {
+                        // TODO(konstin): We should attach the unavailability to the versions we
+                        // rejected in `choose_version` due to it, not generically per package.
```
Comment (zanieb) [inline crates/uv-resolver/src/resolver/mod.rs:369]: Usually these are for packages that are fully unavailable, not a version.

Comment (konstin) [reply crates/uv-resolver/src/resolver/mod.rs:369]: Attached it to the relevant branch

**Comment by konstin** [comment]
The whole plural handling imho makes a lot of effort over just reformulating plural-independent

File: crates/uv/tests/pip_compile.rs [hunk 4]
```diff
@@ -4560,7 +4560,7 @@ fn no_index_requirements_txt() -> Result<()> {
 
     ----- stderr -----
       × No solution found when resolving dependencies:
-      ╰─▶ Because tqdm was not found in the provided package locations and you require tqdm, we can conclude that the requirements are unsatisfiable.
+      ╰─▶ Because all versions of tqdm were not found in the provided package locations and you require tqdm, we can conclude that the requirements are unsatisfiable.
```
Comment (konstin) [inline crates/uv/tests/pip_compile.rs:4563]: Here the formatter methods would need to match on the reason whether it's a whole package or only a version

File: crates/uv/tests/pip_compile.rs [hunk 5]
```diff
@@ -4668,7 +4668,7 @@ fn offline_registry() -> Result<()> {
 
     ----- stderr -----
       × No solution found when resolving dependencies:
-      ╰─▶ Because black==23.10.1 was not found in the cache and you require black==23.10.1, we can conclude that the requirements are unsatisfiable.
+      ╰─▶ Because black was not found in the cache and you require black==23.10.1, we can conclude that the requirements are unsatisfiable.
```
Comment (konstin) [inline crates/uv/tests/pip_compile.rs:4671]: The message improves here because it's the versions of black, not that specific version, that is missing in the cache

File: crates/uv/tests/pip_sync.rs [hunk 6]
```diff
@@ -4658,7 +4658,7 @@ fn require_hashes_registry_valid_hash() -> Result<()> {
 
     ----- stderr -----
       × No solution found when resolving dependencies:
-      ╰─▶ Because example-a-961b4c22==1.0.0 was not found in the package registry and you require example-a-961b4c22==1.0.0, we can conclude that the requirements are unsatisfiable.
+      ╰─▶ Because example-a-961b4c22 was not found in the package registry and you require example-a-961b4c22==1.0.0, we can conclude that the requirements are unsatisfiable.
```
Comment (zanieb) [inline crates/uv/tests/pip_sync.rs:4661]: I'm confused why does this test error? Shouldn't this resolve cc @charliermarsh

**Review by zanieb — APPROVED** [review]
Nice!
